ledger/src/bank_forks.rs

```diff
@@ -334,10 +335,17 @@ impl BankForks {
         Ok(())
     }
 
-    fn prune_non_root(&mut self, root: Slot) {
+    fn prune_non_root(&mut self, root: Slot, largest_confirmed_root: Option<Slot>) {
         let descendants = self.descendants();
-        self.banks
-            .retain(|slot, _| slot == &root || descendants[&root].contains(slot));
+        self.banks.retain(|slot, _| {
+            *slot == root
+                || descendants[&root].contains(slot)
+                || *slot < root && *slot >= largest_confirmed_root.unwrap_or(root)
```

Comment:
I think this third condition you probably want to also add `&& descendants[&slot].contains(root)` because you only want to retain ancestors of the root

Reply:
I think ^ this should also guarantee that any banks `b` `largest_confirmed_root < b < root` so ReplayStage calls `bank_forks.active_banks`, we don't try to replay any slots less than root

Reply:
> I think this third condition you probably want to also add `&& descendants[&slot].contains(root)` because you only want to retain ancestors of the root

Do you mean:
```
let ancestors = self.ancestors();
...
  && ancestors[&slot].contains(root)
```
?

Reply:
@CriesofCarrots no I meant change the last condition from:

`|| *slot < root && *slot >= largest_confirmed_root.unwrap_or(root)` 

to 

`|| *slot < root && *slot >= largest_confirmed_root.unwrap_or(root) && descendants[&slot].contains(root)` 

This way, any remaining banks `b`, where `largest_confirmed_root < b < root` are an ancestor of `root`.

Reply:
Yeah, sorry, I figured it out too slowly :)